Vacation Mode on Alarm Panels

posixx opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
- Implements feature request https://community.home-assistant.io/t/alarm-control-panel-vacation-mode/213872/2
- already discussed in architectual repo: https://github.com/home-assistant/architecture/discussions/500
- new constants are introduced (const.py) following naming of existing alarmpanel constants

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
- Link to dev documentation pull request: https://github.com/home-assistant/developers.home-assistant/pull/758
- Implements https://community.home-assistant.io/t/alarm-control-panel-vacation-mode/213872/2 and includes MQTT Alarmpanel changes

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum
<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io


Proof of working vacation mode:

![image](https://user-images.githubusercontent.com/2280400/88560861-f5416c80-d02e-11ea-9a8b-24a280d75534.png)

![image](https://user-images.githubusercontent.com/2280400/88560901-01c5c500-d02f-11ea-92e6-365467ef4a59.png)

posixx committed:
Introduct vacation mode for alarm panel

posixx committed:
MQTT and Const update

posixx committed:
Merge branch 'dev' into dev

posixx committed:
Update __init__.py

posixx committed:
Update __init__.py

posixx committed:
Update __init__.py

posixx committed:
Update __init__.py

posixx committed:
Update pt.json

posixx committed:
Update pt-BR.json

posixx committed:
Update device_action.py

posixx committed:
Update device_action.py

posixx committed:
Update device_action.py

posixx committed:
MQTT changes ans black fix

posixx committed:
Merge branch 'dev' of https://github.com/posixx/core into dev

posixx committed:
miising payload_arm_vacation

posixx committed:
Delete .gitignore

posixx committed:
missing abbreviation for payload_arm_vacation

posixx committed:
Merge branch 'dev' of https://github.com/posixx/core into dev

posixx committed:
Merge remote-tracking branch 'upstream/dev' into dev

posixx committed:
Revert "Delete .gitignore"

This reverts commit 419e2dafc3f2500bcd857f277547d19971e04fab.

posixx committed:
Merge remote-tracking branch 'upstream/dev' into dev

posixx committed:
Update homeassistant/components/alarm_control_panel/device_trigger.py

Co-authored-by: Nathan Tilley <nathan@tilley.xyz>

posixx committed:
Update .gitignore

Co-authored-by: Nathan Tilley <nathan@tilley.xyz>

posixx committed:
Use constant as supposed

posixx committed:
Revert "Use constant as supposed"

This reverts commit bb238a7737c603cb8609e7af520946ea6b1fba06.

posixx committed:
Add ATTR_SUPPORTED_FEATURES constant as expected

posixx committed:
Add ATTR_SUPPORTED_FEATURES constant as expected

posixx committed:
Merge branch 'dev' of https://github.com/posixx/core into dev

posixx committed:
Update Unit tests

posixx committed:
Unit test adding vacation mode

posixx committed:
payload fix

posixx committed:
condition counter fix

posixx committed:
arm_vacation_calls fix

posixx committed:
tryfix

bachya commented:
This is a pretty large architectural change; it should be first discussed via an issue in [the architecture repo](https://github.com/home-assistant/architecture).

posixx commented:
> This is a pretty large architectural change; it should be first discussed via an issue in [the architecture repo](https://github.com/home-assistant/architecture).

Already extensively discussed: https://github.com/home-assistant/architecture/discussions/500

bachya commented:
@posixx Please make sure that's included in the issue body so that others are aware.

posixx commented:
> @posixx Please make sure that's included in the issue body so that others are aware.

Tnx, done that.

posixx committed:
Revert "tryfix"

This reverts commit c685cb78d12fa825635afefe024b7a1b64ef7ab5.

posixx committed:
supported_features fix

posixx committed:
action value fix

posixx committed:
Some code fixes

posixx committed:
remove pending alarmstate

WHy this is there in the first place is unclear; current core source lacks state also

posixx committed:
fixed typo

MartinHjelmare commented:
Why did you open a new PR instead of continuing the old PR?

https://github.com/home-assistant/core/pull/38284

posixx commented:
> Why did you open a new PR instead of continuing the old PR?
> 
> #38284

I didn't close it, see previous PR. I cannot give up after being almost there:-)

MartinHjelmare commented:
If you wanted to continue you could have just asked to have it reopen. It was only closed 12 hours ago.

posixx commented:
> If you wanted to continue you could have just asked to have it reopen. It was only closed 12 hours ago.

You can read in the previous PR that i asked for reopen. @ntilley905 told me to open a new one. So i did.

ntilley905 commented:
Sorry, that's my fault - I misunderstood the comment on the old PR. I thought you had found typos in the existing code, not the changes you had made. Sorry about that!

posixx commented:
It seems i get some unittest-related errors now:

Error:       raise RuntimeError('Event loop is closed')

are these due to some changed in the build process?

MartinHjelmare commented on homeassistant/components/alarm_control_panel/translations/af.json:
```diff
@@ -1,17 +1,18 @@
-            "pending": "Hangende",
-            "triggered": "Geaktiveer"
-        }
+      "_": {
```
Please don't change translation files.

posixx replied:
Please explain; these files are used by the alarm_control_panel component. Do you have a translation team that takes care of this?

MartinHjelmare replied:
The translation files are synced automatically from Lokalise. The source of truth is `strings.json`.

You can read more here:
https://developers.home-assistant.io/docs/translations

MartinHjelmare commented on homeassistant/components/mqtt/alarm_control_panel.py:
```diff
@@ -11,6 +11,7 @@
     SUPPORT_ALARM_ARM_CUSTOM_BYPASS,
     SUPPORT_ALARM_ARM_HOME,
     SUPPORT_ALARM_ARM_NIGHT,
+    SUPPORT_ALARM_ARM_VACATION,
```
Please don't change any platforms in this PR. That should be done in separate PRs after this is merged.

posixx replied:
So mqtt and manual_mqtt changes need to be reverted and put in later after merge?

MartinHjelmare replied:
Yes, please. 👍

posixx committed:
Revert commits as requested

posixx committed:
Merge remote-tracking branch 'upstream/dev' into dev

posixx committed:
revert mqtt as requested

posixx committed:
revert mqtt test

posixx committed:
Merge remote-tracking branch 'upstream/dev' into dev

posixx commented:
Guys, im unable to update my PR since saturday. When i use github desktop or visual studio on commit i get:

/usr/bin/env: 'python3.8': No such file or directory

I tried everything but it refuses to work. This morning i even restored my system drive to the state of staurday morning. But i still get this error! So i'm starting to suspect the fork itself. Could this be the source of this issue?

posixx committed:
test fixes

posixx commented:
Ok i really need help resolving the test issues. I an unable to find why these constantly fail.

posixx committed:
tryfix tests

posixx committed:
missing arm_vacation in init

posixx committed:
Remove arm_vacation test event

posixx committed:
revert arm_vacation test

posixx committed:
tryfix

posixx committed:
keep trying

posixx committed:
try feature fix

posixx committed:
test fix

posixx committed:
Revert "test fix"

This reverts commit 171162890a0c7ac00a7f9adc58086df699b1a3ea.

posixx committed:
tryfix

posixx commented:
Why this change request isn't merged already is unclear to me. Tried to wakeup multiple times but iit seems noone cares

MartinHjelmare commented on homeassistant/components/alarm_control_panel/const.py:
```diff
@@ -3,12 +3,15 @@
 SUPPORT_ALARM_ARM_NIGHT = 4
-SUPPORT_ALARM_TRIGGER = 8
-SUPPORT_ALARM_ARM_CUSTOM_BYPASS = 16
+SUPPORT_ALARM_ARM_VACATION = 8
```
Let's not change the existing support integers. Just add the new modes 32.

posixx replied:
Fixed

MartinHjelmare commented on tests/components/alarm_control_panel/common.py:
```diff
@@ -61,6 +62,19 @@ async def async_alarm_arm_night(hass, code=None, entity_id=ENTITY_MATCH_ALL):
     await hass.services.async_call(DOMAIN, SERVICE_ALARM_ARM_NIGHT, data, blocking=True)
 
 
+async def async_alarm_arm_vacation(hass, code=None, entity_id=ENTITY_MATCH_ALL):
```
Please don't add this helper. We're trying to remove these helpers. Just call the service directly.

posixx replied:
If i remove this it breaks the tests. If this needs to be removed please provide replacement code as i don't have the knowlegde how to solve

posixx replied:
fixed with correction support integers

MartinHjelmare replied:
Where is this used?

posixx replied:
helper removed tnx to @KJonline

MartinHjelmare commented on tests/components/alarm_control_panel/test_device_action.py:
```diff
@@ -297,6 +320,13 @@ async def test_action(hass):
+    await hass.async_block_till_done()
+    assert (
+        hass.states.get("alarm_control_panel.alarm_no_arm_code").state
+        != STATE_ALARM_ARMED_VACATION
```
What is the goal of this test?

posixx replied:
This was added because with "==" build fails everytime. Please provide a solution if you know how to solve

posixx replied:
this one keeps giving troubles with using "==". See https://github.com/home-assistant/core/actions/runs/661327554

posixx replied:
tried everything, this is not related to wrong coding. i give up.

KJonline replied:
@posixx ill take a look and see if I can help with tests issue.

posixx committed:
Merge remote-tracking branch 'upstream/dev' into dev

posixx committed:
proposed changes

posixx committed:
Removed vacation service

posixx committed:
match new support integers

posixx committed:
fix tests

posixx committed:
revert becase of build error

THids causer assert error:

arm_night == arm_vacation

posixx committed:
trigger test fix

posixx committed:
tryfix action test (changed order)

Commit:
Update test configuration.

posixx committed:
Merge pull request #1 from KJonline/dev

Alarm Vacation Mode - Update test configuration.

posixx committed:
Merge remote-tracking branch 'upstream/dev' into dev

posixx committed:
add missing vacation test

MartinHjelmare commented on tests/testing_config/custom_components/test/alarm_control_panel.py:
```diff
@@ -96,6 +99,11 @@ def alarm_arm_night(self, code=None):
+    def alarm_arm_vacation(self, code=None):
+        """Send arm night command."""
+        self._state = STATE_ALARM_ARMED_VACATION
+        self.async_write_ha_state()
```
```suggestion
        self.schedule_update_ha_state()
```

posixx committed:
Update tests/testing_config/custom_components/test/alarm_control_panel.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

posixx commented:
is it possible someon takes over ownership of this PR? I am committing things to solve but have no idea what i am doing... I think i make things worse..

Commit:
Merge pull request #1 from posixx/dev

Update from fork

Commit:
Fix Tests

posixx committed:
Merge pull request #2 from KJonline/dev

Fix tests for vacation mode

Commit:
remove unused helper

posixx committed:
Merge pull request #3 from KJonline/dev

Remove alarm helper

emontnemery committed:
Merge branch 'dev' into dev

emontnemery reviewed (approved):
Looks good, thanks a lot @posixx and @KJonline 👍

frenck commented on homeassistant/components/alarm_control_panel/services.yaml:
```diff
@@ -70,6 +70,16 @@ alarm_arm_night:
       selector:
         text:
 
+alarm_arm_vacation:
+  description: Send the alarm the command for arm vacation.
+  fields:
+    entity_id:
+      description: Name of alarm control panel to arm vacation.
+      example: "alarm_control_panel.downstairs"
+    code:
+      description: An optional code to arm vacation the alarm control panel with.
+      example: "1234"
```
```suggestion
alarm_arm_vacation:
  name: Arm vacation
  description: Send the alarm the command for arm vacation.
  target:
    entity:
      domain: alarm_control_panel
  fields:
    code:
      name: Code
      description: An optional code to arm vacation the alarm control panel with.
      example: "1234"
      selector:
        text:
```

posixx committed:
Update homeassistant/components/alarm_control_panel/services.yaml

Co-authored-by: Franck Nijhof <frenck@frenck.nl>
